include/subsystems/odometry/odometry_nwheel.h

```diff
@@ -0,0 +1,271 @@
+    static timer tmr;
+
+    double speed_local = 0;
+    double accel_local = 0;
+    double ang_speed_local = 0;
+    double ang_accel_local = 0;
+    bool update_vel_accel = tmr.time(sec) > 0.1;
+
+    // This loop runs too fast. Only check at LEAST every 1/10th sec
+    if (update_vel_accel) {
+      // Calculate robot velocity
+      speed_local = pos_diff(updated_pos, last_pos) / tmr.time(sec);
+
+      // Calculate robot acceleration
+      accel_local = (speed_local - last_speed) / tmr.time(sec);
+
+      // Calculate robot angular velocity (deg/sec)
+      ang_speed_local = smallest_angle(updated_pos.rot, last_pos.rot) / tmr.time(sec);
+
+      // Calculate robot angular acceleration (deg/sec^2)
+      ang_accel_local = (ang_speed_local - last_ang_speed) / tmr.time(sec);
+
+      tmr.reset();
+      last_pos = updated_pos;
+      last_speed = speed_local;
+      last_ang_speed = ang_speed_local;
+    }
+
+    this->current_pos = updated_pos;
+    if (update_vel_accel) {
+      this->speed = speed_local;
+      this->accel = accel_local;
+      this->ang_speed_deg = ang_speed_local;
+      this->ang_accel_deg = ang_accel_local;
+    }
+
+    if (imu != nullptr) {
+      old_heading = angle;
+    }
+
+    return current_pos;
+  }
+
+  /**
+   * Resets the position and rotational data to the input.
+   */
+  void set_position(const pose_t &newpos) override {
+    mut.lock();
+    rotation_offset = newpos.rot - (current_pos.rot - rotation_offset);
+    mut.unlock();
+
+    OdometryBase::set_position(newpos);
+  }
+
+private:
+  /**
+   * Applies a twist (pose delta) to a pose by including first order dynamics of heading.
+   * Can be thought of as applying a twist as following an arc rather than a straight line.
+   *
+   * https://file.tavsys.net/control/controls-engineering-in-frc.pdf#section.10.2
+   *
+   * @param old_pose  The pose to which the twist will be applied
+   * @param twist     The twist, represents a pose delta
+   */
+  pose_t pose_exponential(const Eigen::Vector3d old_pose, const Eigen::Vector3d twist) {
+    double dtheta = twist(2);
+
+    double sinTheta = sin(dtheta);
+    double cosTheta = cos(dtheta);
+
+    double sinOldTheta = sin(old_pose(2));
+    double cosOldTheta = cos(old_pose(2));
+
+    Eigen::Matrix3d rotation{{cosOldTheta, -sinOldTheta, 0}, {sinOldTheta, cosOldTheta, 0}, {0, 0, 1}};
+    Eigen::Matrix3d integrated_rotation;
+
+    // when the angle change is very small, we use a taylor series to approximate
+    if (abs(dtheta) < 1e-9) {
+      integrated_rotation.row(0) << 1.0 - ((dtheta * dtheta) / 6.0), -(dtheta / 2.0), 0;
+      integrated_rotation.row(1) << (dtheta / 2.0), 1.0 - ((dtheta * dtheta) / 6.0), 0;
+      integrated_rotation.row(2) << 0, 0, 1;
+    } else {
+      integrated_rotation.row(0) << sinTheta / dtheta, (cosTheta - 1.0) / dtheta, 0;
+      integrated_rotation.row(1) << (1 - cosTheta) / dtheta, sinTheta / dtheta, 0;
+      integrated_rotation.row(2) << 0, 0, 1;
+    }
+
```

Comment:
Probably worth adding a comment on where the constants came from. Such as the 6 the 2